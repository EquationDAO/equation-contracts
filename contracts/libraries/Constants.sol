// SPDX-License-Identifier: BUSL-1.1
pragma solidity ^0.8.0;

library Constants {
    uint32 internal constant BASIS_POINTS_DIVISOR = 100_000_000;

    uint16 internal constant ADJUST_FUNDING_RATE_INTERVAL = 1 hours;
    uint16 internal constant SAMPLE_PREMIUM_RATE_INTERVAL = 5 seconds;
    uint16 internal constant REQUIRED_SAMPLE_COUNT = ADJUST_FUNDING_RATE_INTERVAL / SAMPLE_PREMIUM_RATE_INTERVAL;
    /// @dev 8 * (1+2+3+...+720) = 8 * ((1+720) * 720 / 2) = 8 * 259560
    uint32 internal constant PREMIUM_RATE_AVG_DENOMINATOR = 8 * 259560;
    /// @dev RoundingUp(50000 / 8 * Q96 / BASIS_POINTS_DIVISOR) = 4951760157141521099596497
    int256 internal constant PREMIUM_RATE_CLAMP_BOUNDARY_X96 = 4951760157141521099596497; // 0.05% / 8

    uint8 internal constant VERTEX_NUM = 7;
    uint8 internal constant LATEST_VERTEX = VERTEX_NUM - 1;

    uint64 internal constant RISK_BUFFER_FUND_LOCK_PERIOD = 90 days;

    uint256 internal constant Q64 = 1 << 64;
    uint256 internal constant Q96 = 1 << 96;

    bytes32 internal constant ROLE_POSITION_LIQUIDATOR = keccak256("ROLE_POSITION_LIQUIDATOR");
    bytes32 internal constant ROLE_LIQUIDITY_POSITION_LIQUIDATOR = keccak256("ROLE_LIQUIDITY_POSITION_LIQUIDATOR");
}
