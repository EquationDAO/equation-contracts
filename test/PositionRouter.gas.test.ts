import {loadFixture} from "@nomicfoundation/hardhat-network-helpers";
import {ethers} from "hardhat";

import {SIDE_LONG, SIDE_SHORT} from "./shared/Constants";
import {expectSnapshotGasCost} from "./shared/snapshotGasCost";

describe("PositionRouter gas tests", function () {
    const defaultMinExecutionFee = 3000;

    async function deployFixture() {
        const [owner, otherAccount1, otherAccount2] = await ethers.getSigners();
        const ERC20Test = await ethers.getContractFactory("ERC20Test");
        const USDC = await ERC20Test.connect(otherAccount1).deploy("USDC", "USDC", 6, 100_000_000n);
        const ETH = await ERC20Test.connect(otherAccount1).deploy("ETHER", "ETH", 18, 100_000_000);
        await USDC.deployed();
        await ETH.deployed();

        const Pool = await ethers.getContractFactory("MockPool");
        const pool = await Pool.deploy(USDC.address, ETH.address);
        await pool.deployed();

        const Router = await ethers.getContractFactory("MockRouter");
        const router = await Router.deploy();
        await router.deployed();

        // router can transfer owner's USDC
        await USDC.connect(otherAccount1).approve(router.address, 100_000_000n);

        const PositionRouter = await ethers.getContractFactory("PositionRouter");
        const positionRouter = await PositionRouter.deploy(USDC.address, router.address, defaultMinExecutionFee);
        await positionRouter.deployed();

        return {
            owner,
            otherAccount1,
            otherAccount2,
            router,
            positionRouter,
            USDC,
            ETH,
            pool,
        };
    }

    it("createIncreasePosition", async () => {
        const {positionRouter, pool, otherAccount1} = await loadFixture(deployFixture);
        await expectSnapshotGasCost(
            positionRouter.connect(otherAccount1).createIncreasePosition(pool.address, SIDE_LONG, 100n, 100n, 100n, {
                value: 3000,
            })
        );
    });

    it("executeIncreasePosition", async () => {
        const {positionRouter, pool, otherAccount1, router} = await loadFixture(deployFixture);
        await positionRouter.updateDelayValues(0n, 0n, 600n);
        await positionRouter
            .connect(otherAccount1)
            .createIncreasePosition(pool.address, SIDE_SHORT, 100n, 100n, 1790n, {
                value: 3000,
            });
        await router.setTradePriceX96(1795n);
        await expectSnapshotGasCost(
            positionRouter.connect(otherAccount1).executeIncreasePosition(0n, otherAccount1.address)
        );
    });
});
