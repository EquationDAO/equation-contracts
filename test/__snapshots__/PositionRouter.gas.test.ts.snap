// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PositionRouter gas tests createIncreasePosition 1`] = `213428`;

exports[`PositionRouter gas tests executeIncreasePosition 1`] = `87900`;
